File: packages/db/src/session-store.ts
import type { SessionKey, SessionStoreEntry } from '@anthropic-ai/claude-agent-sdk';
import { withSpan } from '@chemclaw2/observability';
import { db } from './client';
import { agentSessions } from './schema/sessions';
import { eq, and, sql, max } from 'drizzle-orm';

// Caps protect against (a) advisory-lock keys derived from pathological inputs
// and (b) OOM when the SDK keeps appending forever. Realistic sessions stay
// well under both limits.
const MAX_KEY_PART_LEN = 256;
const MAX_APPEND_ENTRIES = 100;
const MAX_ENTRY_SERIALIZED_BYTES = 1_000_000;

function assertKeyComponent(name: string, value: string): void {
  if (value.length > MAX_KEY_PART_LEN) {
    throw new Error(`session-store: ${name} exceeds ${MAX_KEY_PART_LEN} chars`);
  }
}

/**
 * Returns a session store that forces every key to use the supplied projectKey,
 * preventing one user's sessions from being accessible via another user's context.
 * The SDK derives projectKey from cwd by default; this wrapper overrides that for
 * multi-user server deployments where cwd is shared across all requests.
 */
export function scopedSessionStore(projectKey: string) {
  const scoped = (key: SessionKey): SessionKey => ({ ...key, projectKey });
  return {
    append: (key: SessionKey, entries: SessionStoreEntry[]) =>
      postgresSessionStore.append(scoped(key), entries),
    load: (key: SessionKey) => postgresSessionStore.load(scoped(key)),
    listSessions: () => postgresSessionStore.listSessions(projectKey),
    delete: (key: SessionKey) => postgresSessionStore.delete(scoped(key)),
    listSubkeys: (key: { projectKey: string; sessionId: string }) =>
      postgresSessionStore.listSubkeys({ ...key, projectKey }),
  };
}

export const postgresSessionStore = {
  async append(key: SessionKey, entries: SessionStoreEntry[]): Promise<void> {
    const subpath = key.subpath ?? '';
    assertKeyComponent('projectKey', key.projectKey);
    assertKeyComponent('sessionId', key.sessionId);
    assertKeyComponent('subpath', subpath);
    if (entries.length > MAX_APPEND_ENTRIES) {
      throw new Error(`session-store: refusing to append ${entries.length} entries (max ${MAX_APPEND_ENTRIES})`);
    }
    const serializedSize = Buffer.byteLength(JSON.stringify(entries), 'utf8');
    if (serializedSize > MAX_ENTRY_SERIALIZED_BYTES) {
      throw new Error(`session-store: entries serialize to ${serializedSize} bytes (max ${MAX_ENTRY_SERIALIZED_BYTES})`);
    }
    const now = Date.now();
    // Serialize concurrent appends with a transaction-scoped advisory lock.
    // The two-arg form composes 64 bits from two 32-bit hashes — collisions
    // between independent (projectKey, sessionId+subpath) pairs only occur if
    // BOTH halves hash-collide, vs. ~1e-9 per single-key hashtext alone.
    //
    // Span carries the advisory-lock + insert path so contention spikes (the
    // common cause of agent-stream latency tails) get attributed correctly.
    await withSpan(
      'session_store.append',
      { entries_count: entries.length, serialized_bytes: serializedSize },
      async () => {
        await db.transaction(async (tx) => {
          await tx.execute(sql`
            SELECT pg_advisory_xact_lock(
              hashtext(${key.projectKey}),
              hashtext(${key.sessionId} || '::' || ${subpath})
            )
          `);
          await tx
            .insert(agentSessions)
            .values({
              projectKey: key.projectKey,
              sessionId: key.sessionId,
              subpath,
              entries,
              mtime: now,
            })
            .onConflictDoUpdate({
              target: [agentSessions.projectKey, agentSessions.sessionId, agentSessions.subpath],
              set: {
                entries: sql`${agentSessions.entries} || excluded.entries`,
                mtime: sql`GREATEST(${agentSessions.mtime}, excluded.mtime)`,
              },
            });
        });
      },
    );
  },

  async load(key: SessionKey): Promise<SessionStoreEntry[] | null> {
    const subpath = key.subpath ?? '';
    const row = await db.query.agentSessions.findFirst({
      where: and(
        eq(agentSessions.projectKey, key.projectKey),
        eq(agentSessions.sessionId, key.sessionId),
        eq(agentSessions.subpath, subpath),
      ),
    });
    return row ? (row.entries as SessionStoreEntry[]) : null;
  },

  async listSessions(projectKey: string): Promise<Array<{ sessionId: string; mtime: number }>> {
    const rows = await db
      .select({
        sessionId: agentSessions.sessionId,
        mtime: max(agentSessions.mtime).as('mtime'),
      })
      .from(agentSessions)
      .where(eq(agentSessions.projectKey, projectKey))
      .groupBy(agentSessions.sessionId);
    return rows
      .filter((r) => r.mtime !== null)
      .map((r) => ({ sessionId: r.sessionId, mtime: r.mtime as number }));
  },

  async delete(key: SessionKey): Promise<void> {
    const subpath = key.subpath ?? '';
    if (subpath === '') {
      // Delete main key + all subkeys for this session
      await db
        .delete(agentSessions)
        .where(
          and(
            eq(agentSessions.projectKey, key.projectKey),
            eq(agentSessions.sessionId, key.sessionId),
          ),
        );
    } else {
      await db
        .delete(agentSessions)
        .where(
          and(
            eq(agentSessions.projectKey, key.projectKey),
            eq(agentSessions.sessionId, key.sessionId),
            eq(agentSessions.subpath, subpath),
          ),
        );
    }
  },

  async listSubkeys(key: { projectKey: string; sessionId: string }): Promise<string[]> {
    const rows = await db
      .select({ subpath: agentSessions.subpath })
      .from(agentSessions)
      .where(
        and(
          eq(agentSessions.projectKey, key.projectKey),
          eq(agentSessions.sessionId, key.sessionId),
        ),
      );
    return rows.map((r) => r.subpath).filter((s) => s !== '');
  },
} satisfies {
  append(key: SessionKey, entries: SessionStoreEntry[]): Promise<void>;
  load(key: SessionKey): Promise<SessionStoreEntry[] | null>;
  listSessions(projectKey: string): Promise<Array<{ sessionId: string; mtime: number }>>;
  delete(key: SessionKey): Promise<void>;
  listSubkeys(key: { projectKey: string; sessionId: string }): Promise<string[]>;
};

File: api/db/queries/session_store.py
"""Postgres-backed SessionStore for the Python Agent SDK.

Mirrors the TypeScript `postgresSessionStore` in packages/db/src/session-store.ts.
Uses the same `agent_sessions` table with advisory locks for concurrent-safe appends.
"""
from __future__ import annotations

import logging
from typing import Any

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

logger = logging.getLogger(__name__)

MAX_KEY_PART_LEN = 256
MAX_APPEND_ENTRIES = 100
MAX_ENTRY_SERIALIZED_BYTES = 1_000_000


def _assert_key_component(name: str, value: str) -> None:
    if len(value) > MAX_KEY_PART_LEN:
        raise ValueError(f"session-store: {name} exceeds {MAX_KEY_PART_LEN} chars")


class PostgresSessionStore:
    """Implements the claude_agent_sdk.types.SessionStore protocol backed by Postgres."""

    def __init__(self, session_factory: async_sessionmaker[AsyncSession]) -> None:
        self._factory = session_factory

    async def append(
        self,
        key: dict[str, Any],  # SessionKey TypedDict
        entries: list[dict[str, Any]],
    ) -> None:
        project_key = key["project_key"]
        session_id = key["session_id"]
        subpath = key.get("subpath") or ""
        _assert_key_component("project_key", project_key)
        _assert_key_component("session_id", session_id)
        _assert_key_component("subpath", subpath)

        if len(entries) > MAX_APPEND_ENTRIES:
            raise ValueError(f"session-store: refusing to append {len(entries)} entries (max {MAX_APPEND_ENTRIES})")

        import json
        serialized = json.dumps(entries).encode()
        if len(serialized) > MAX_ENTRY_SERIALIZED_BYTES:
            raise ValueError(
                f"session-store: entries serialize to {len(serialized)} bytes "
                f"(max {MAX_ENTRY_SERIALIZED_BYTES})"
            )

        import time
        mtime = int(time.time() * 1000)

        async with self._factory() as db:
            async with db.begin():
                # Advisory lock: two 32-bit hashes combined → 64-bit lock key.
                # Prevents concurrent appends from interleaving entries.
                await db.execute(text("""
                    SELECT pg_advisory_xact_lock(
                        hashtext(:project_key),
                        hashtext(:session_id || '::' || :subpath)
                    )
                """), {"project_key": project_key, "session_id": session_id, "subpath": subpath})

                await db.execute(text("""
                    INSERT INTO agent_sessions (project_key, session_id, subpath, entries, mtime)
                    VALUES (:project_key, :session_id, :subpath, CAST(:entries AS jsonb), :mtime)
                    ON CONFLICT (project_key, session_id, subpath) DO UPDATE
                        SET entries   = agent_sessions.entries || EXCLUDED.entries,
                            mtime     = EXCLUDED.mtime,
                            insert_seq = DEFAULT
                """), {
                    "project_key": project_key,
                    "session_id": session_id,
                    "subpath": subpath,
                    "entries": serialized.decode(),
                    "mtime": mtime,
                })

    async def load(
        self,
        key: dict[str, Any],
    ) -> list[dict[str, Any]]:
        project_key = key["project_key"]
        session_id = key["session_id"]
        subpath = key.get("subpath") or ""

        async with self._factory() as db:
            result = await db.execute(text("""
                SELECT entries
                FROM agent_sessions
                WHERE project_key = :project_key
                  AND session_id  = :session_id
                  AND subpath     = :subpath
                ORDER BY insert_seq
            """), {"project_key": project_key, "session_id": session_id, "subpath": subpath})
            rows = result.fetchall()
            out = []
            for (entries,) in rows:
                if isinstance(entries, list):
                    out.extend(entries)
            return out

    async def list_sessions(self, project_key: str) -> list[str]:
        async with self._factory() as db:
            result = await db.execute(text("""
                SELECT DISTINCT session_id
                FROM agent_sessions
                WHERE project_key = :project_key AND subpath = ''
                ORDER BY session_id
            """), {"project_key": project_key})
            return [r.session_id for r in result]

    async def delete(self, key: dict[str, Any]) -> None:
        project_key = key["project_key"]
        session_id = key["session_id"]
        subpath = key.get("subpath") or ""
        async with self._factory() as db:
            await db.execute(text("""
                DELETE FROM agent_sessions
                WHERE project_key = :project_key
                  AND session_id  = :session_id
                  AND (:subpath = '' OR subpath = :subpath)
            """), {"project_key": project_key, "session_id": session_id, "subpath": subpath})
            await db.commit()


def scoped_session_store(
    session_factory: async_sessionmaker[AsyncSession],
    project_key: str,
) -> "ScopedPostgresSessionStore":
    """Returns a store that forces every key to use the given project_key.
    Mirrors TypeScript's scopedSessionStore() for multi-tenant isolation.
    """
    return ScopedPostgresSessionStore(session_factory, project_key)


class ScopedPostgresSessionStore(PostgresSessionStore):
    def __init__(
        self,
        session_factory: async_sessionmaker[AsyncSession],
        forced_project_key: str,
    ) -> None:
        super().__init__(session_factory)
        self._forced_project_key = forced_project_key

    def _scoped(self, key: dict[str, Any]) -> dict[str, Any]:
        return {**key, "project_key": self._forced_project_key}

    async def append(self, key: dict[str, Any], entries: list[dict[str, Any]]) -> None:
        await super().append(self._scoped(key), entries)

    async def load(self, key: dict[str, Any]) -> list[dict[str, Any]]:
        return await super().load(self._scoped(key))

    async def delete(self, key: dict[str, Any]) -> None:
        await super().delete(self._scoped(key))

    async def list_sessions(self, project_key: str) -> list[str]:
        return await super().list_sessions(self._forced_project_key)
